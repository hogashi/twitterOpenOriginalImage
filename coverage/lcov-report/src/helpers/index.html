<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/helpers</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/helpers
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.1% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>90/219</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">26.32% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>30/114</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.3% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>19/46</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.47% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>90/217</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="ButtonSetter.ts"><a href="ButtonSetter.ts.html">ButtonSetter.ts</a></td>
	<td data-value="25" class="pic low"><div class="chart"><div class="cover-fill" style="width: 25%;"></div><div class="cover-empty" style="width:75%;"></div></div></td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="104" class="abs low">26/104</td>
	<td data-value="7.27" class="pct low">7.27%</td>
	<td data-value="55" class="abs low">4/55</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="24" class="abs low">8/24</td>
	<td data-value="25.49" class="pct low">25.49%</td>
	<td data-value="102" class="abs low">26/102</td>
	</tr>

<tr>
	<td class="file low" data-value="ButtonSetterTweetDeck.ts"><a href="ButtonSetterTweetDeck.ts.html">ButtonSetterTweetDeck.ts</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="51" class="abs low">0/51</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="33" class="abs low">0/33</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="51" class="abs low">0/51</td>
	</tr>

<tr>
	<td class="file high" data-value="ButtonSetters.ts"><a href="ButtonSetters.ts.html">ButtonSetters.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	</tr>

<tr>
	<td class="file high" data-value="Constants.ts"><a href="Constants.ts.html">Constants.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="14" class="abs high">14/14</td>
	</tr>

<tr>
	<td class="file high" data-value="Utils.ts"><a href="Utils.ts.html">Utils.ts</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="47" class="abs high">47/47</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="26" class="abs high">26/26</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="11" class="abs high">11/11</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="47" class="abs high">47/47</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Sun Jul 07 2019 20:49:52 GMT+0900 (GMT+09:00)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
