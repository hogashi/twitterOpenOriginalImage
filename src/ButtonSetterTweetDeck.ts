import { ButtonSetterType } from './ButtonSetter';
import { SHOW_ON_TWEETDECK_TIMELINE, isFalse, SHOW_ON_TWEETDECK_TWEET_DETAIL, Options } from './constants';
import { printException, setStyle, onOriginalButtonClick } from './utils';

/**
 * tweetdeck.twitter.comでボタンを設置するクラス
 */
export class ButtonSetterTweetDeck implements ButtonSetterType {
  // タイムラインにボタン表示
  public setButtonOnTimeline(currentOptions: Options): void {
    // タイムラインにボタン表示する設定がされているときだけ実行する
    // - isTrue か 設定なし のとき ON
    // - isFalse のとき OFF
    if (!(currentOptions[SHOW_ON_TWEETDECK_TIMELINE] !== isFalse)) {
      return;
    }
    // if タイムラインのツイートを取得できたら
    // is-actionable: タイムラインのみ
    const tweets = document.getElementsByClassName('js-stream-item is-actionable') as HTMLCollectionOf<HTMLElement>;
    if (!tweets.length) {
      return;
    }
    const className = 'tooi-button-container-tweetdeck-timeline';
    // 各ツイートに対して
    Array.from(tweets).forEach((tweet) => {
      if (
        !tweet.getElementsByClassName('js-media-image-link').length ||
        tweet.getElementsByClassName('is-video').length ||
        tweet.getElementsByClassName('is-gif').length ||
        tweet.getElementsByClassName(className).length
      ) {
        // メディアツイートでない
        // または メディアが画像でない(動画/GIF)
        // または すでにボタンをおいてあるとき
        // 何もしない
        return;
      }

      const target = tweet.querySelector('footer');
      if (!target) {
        // ボタンを置く場所がないとき
        // 何もしない
        printException('no target');
        return;
      }

      const getImgSrcs = (): string[] => {
        return Array.from(tweet.getElementsByClassName('js-media-image-link'))
          .map((element) => {
            const urlstr = this.getBackgroundImageUrl(element as HTMLElement);
            // filter で string[] にするためにここで string[] にする……
            return urlstr ? urlstr : '';
          })
          .filter((urlstr) => urlstr !== '');
      };
      this.setButton({
        className,
        getImgSrcs,
        target,
      });
    });
  }

  // ツイート詳細にボタン表示
  public setButtonOnTweetDetail(currentOptions: Options): void {
    // ツイート詳細にボタン表示する設定がされているときだけ実行する
    // - isTrue か 設定なし のとき ON
    // - isFalse のとき OFF
    if (!(currentOptions[SHOW_ON_TWEETDECK_TWEET_DETAIL] !== isFalse)) {
      return;
    }
    // if ツイート詳細を取得できたら
    const tweets = document.getElementsByClassName('js-tweet-detail') as HTMLCollectionOf<HTMLElement>;
    if (!tweets.length) {
      return;
    }
    const className = 'tooi-button-container-tweetdeck-detail';
    // 各ツイートに対して
    Array.from(tweets).forEach((tweet) => {
      if (
        !(tweet.getElementsByClassName('media-img').length || tweet.getElementsByClassName('media-image').length) ||
        tweet.getElementsByClassName(className).length
      ) {
        // メディアツイートでない (画像のタグが取得できない)
        // または すでにボタンをおいてあるとき
        // 何もしない
        return;
      }
      const target = tweet.querySelector('footer');
      if (!target) {
        // ボタンを置く場所がないとき
        // 何もしない
        printException('no target');
        return;
      }

      const getImgSrcs = (): string[] => {
        if (tweet.getElementsByClassName('media-img').length !== 0) {
          return [(tweet.getElementsByClassName('media-img')[0] as HTMLImageElement).src];
        } else {
          return Array.from(tweet.getElementsByClassName('media-image'))
            .map((element) => {
              const urlstr = this.getBackgroundImageUrl(element as HTMLElement);
              // filter で string[] にするためにここで string[] にする……
              return urlstr ? urlstr : '';
            })
            .filter((urlstr) => urlstr !== '');
        }
      };

      this.setButton({
        className,
        getImgSrcs,
        target,
      });
    });
  }

  private setButton({
    className,
    getImgSrcs,
    target,
  }: {
    className: string;
    getImgSrcs: () => string[];
    target: HTMLElement;
  }): void {
    // 枠線の色は'Original'と同じく'.txt-mute'の色を使うので取得する
    const txtMute = document.querySelector('.txt-mute');
    const borderColor = txtMute ? window.getComputedStyle(txtMute).color : '#697b8c';
    // ボタンのスタイル設定
    const style = {
      border: `1px solid ${borderColor}`,
      'border-radius': '2px',
      display: 'inline-block',
      'font-size': '0.75em',
      'margin-top': '5px',
      padding: '1px 1px 0',
      'line-height': '1.5em',
      cursor: 'pointer',
    };

    /* つくるDOMは以下 */
    /*
      <a
        className={className}
        style={style}
        onClick={(e) => {
          onOriginalButtonClick(e, imgSrcs);
        }}
      >
        Original
      </a>
    */

    // tweetdeckのツイート右上の時刻などに使われているclassを使う
    // 設置の有無の判別用にclassNameを付与する
    const button = document.createElement('a');
    button.className = `pull-left margin-txs txt-mute ${className}`;
    setStyle(button, style);
    button.addEventListener('click', (e) => {
      onOriginalButtonClick(e, getImgSrcs());
    });
    button.insertAdjacentHTML('beforeend', 'Original');

    target.appendChild(button);
  }

  private getBackgroundImageUrl(element: HTMLElement): string | null {
    if (element.style.backgroundImage) {
      return element.style.backgroundImage.replace(/url\("?([^"]*)"?\)/, '$1');
    }
    return null;
  }
}
