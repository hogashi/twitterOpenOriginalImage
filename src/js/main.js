/* main.js */
// https://twitter.com/* で実行される

tooiInit(setButtons);

// キー押下時
document.addEventListener('keydown', function(e) {
  // if [RETURN(ENTER)]キーなら
  if (
    e.key === 'Enter' &&
    !e.ctrlKey &&
    !e.metaKey &&
    !e.altKey &&
    !e.shiftKey
  ) {
    // ツイート詳細にボタン表示する設定がされていたら
    // かつ ツイート入力ボックスがアクティブでないなら
    if (
      options[OPEN_WITH_KEY_PRESS] !== 'isfalse' &&
      !document.activeElement.className.match(/rich-editor/)
    ) {
      openFromTweetDetail(e);
    }
  }
});

// ボタンを置く
function setButtons() {
  // console.log('setButtons: ' + options['SHOW_ON_TIMELINE'] + ' ' + options['SHOW_ON_TWEET_DETAIL'] + ' ' + options['OPEN_WITH_KEY_PRESS']) // debug
  // if タイムラインにボタン表示する設定がされていたら
  if (options[SHOW_ON_TIMELINE] !== 'isfalse') {
    setButtonOnTimeline();
  }
  // if ツイート詳細にボタン表示する設定がされていたら
  if (options[SHOW_ON_TWEET_DETAIL] !== 'isfalse') {
    setButtonOnTweetDetail();
  }
} // setButtons end

// タイムラインにボタン表示
function setButtonOnTimeline() {
  const tweets = document.getElementsByClassName('js-stream-tweet');
  if (!tweets.length) {
    return;
  }
  // 各ツイートに対して
  Array.from(tweets).forEach(tweet => {
    // if 画像ツイート
    // かつ まだ処理を行っていないなら
    if (
      !!tweet.getElementsByClassName('AdaptiveMedia-container')[0] &&
      !!tweet
        .getElementsByClassName('AdaptiveMedia-container')[0]
        .getElementsByTagName('img')[0] &&
      !tweet.getElementsByClassName('tooiDivTimeline')[0]
    ) {
      let actionList, parentDiv, origButton;
      // ボタンを設置
      // 操作ボタンの外側は様式にあわせる
      actionList = tweet.getElementsByClassName('ProfileTweet-actionList')[0];
      parentDiv = document.createElement('div');
      // parentDiv.id = '' + tweet.id
      parentDiv.className = 'ProfileTweet-action tooiDivTimeline';
      actionList.appendChild(parentDiv);
      // Originalボタン
      origButton = document.createElement('input');
      tweet
        .getElementsByClassName('tooiDivTimeline')[0]
        .appendChild(origButton);
      origButton.type = 'button';
      origButton.value = 'Original';
      origButton.style.width = '70px';
      origButton.style.fontSize = '13px';
      origButton.style.color = '#000000';
      origButton.addEventListener('click', openFromTimeline);
    }
  });
} // setButtonOnTimeline end

// ツイート詳細にボタン表示
function setButtonOnTweetDetail() {
  if (
    !document.getElementsByClassName('permalink-tweet-container')[0] ||
    !document
      .getElementsByClassName('permalink-tweet-container')[0]
      .getElementsByClassName('AdaptiveMedia-photoContainer')[0] ||
    document.getElementById('tooiInputDetailpage')
  ) {
    // ツイート詳細ページでない、または、メインツイートが画像ツイートでないとき
    // または、すでに処理を行ってあるとき
    // 何もしない
    return;
  }
  let actionList, parentDiv, origButton;
  // Originalボタンの親の親となる枠
  actionList = document
    .getElementsByClassName('permalink-tweet-container')[0]
    .getElementsByClassName('ProfileTweet-actionList')[0];
  // Originalボタンの親となるdiv
  parentDiv = document.createElement('div');
  parentDiv.id = 'tooiDivDetailpage';
  parentDiv.className = 'ProfileTweet-action';
  actionList.appendChild(parentDiv);
  // Originalボタン(input type='button')
  origButton = document.createElement('input');
  document.getElementById('tooiDivDetailpage').appendChild(origButton);
  origButton.id = 'tooiInputDetailpage';
  origButton.type = 'button';
  origButton.value = 'Original';
  origButton.style.width = '70px';
  origButton.style.fontSize = '13px';
  origButton.addEventListener('click', openFromTweetDetail);
} // setButtonOnTweetDetail end

// タイムラインから画像を新しいタブに開く
function openFromTimeline(e) {
  // ツイートの画像の親エレメントを取得するためにその親まで遡る
  const parentNode = this.parentNode.parentNode.parentNode.parentNode;
  // if 上述のエレメントが取得できたら
  if (parentNode.getElementsByClassName('AdaptiveMedia-container')[0]) {
    // イベント(MouseEvent)による既定の動作をキャンセル
    e.preventDefault();
    // イベント(MouseEvent)の親要素への伝播を停止
    e.stopPropagation();
    openImagesInNewTab(
      Array.from(
        parentNode.getElementsByClassName('AdaptiveMedia-photoContainer')
      ).map(element => element.getElementsByTagName('img')[0].src)
    );
  } else {
    printException('no image elements on timeline');
  }
} // openFromTimeline end

// ツイート詳細から画像を新しいタブに開く
function openFromTweetDetail(e) {
  // .permalink-tweet-container: ツイート詳細ページのメインツイート
  // .AdaptiveMedia-photoContainer: 画像の親エレメント
  if (!!document.getElementsByClassName('permalink-tweet-container')[0]) {
    // イベント(MouseEvent)による既定の動作をキャンセル
    e.preventDefault();
    // イベント(MouseEvent)の親要素への伝播を停止
    e.stopPropagation();
    openImagesInNewTab(
      Array.from(
        document
          .getElementsByClassName('permalink-tweet-container')[0]
          .getElementsByClassName('AdaptiveMedia-photoContainer')
      ).map(element => element.getElementsByTagName('img')[0].src)
    );
  } else {
    printException('no tweet elements on tweet detail');
  }
} // openFromTweetDetail end
