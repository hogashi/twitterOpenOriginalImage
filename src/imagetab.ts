/* imagetab.ts */
// https://pbs.twimg.com/* で実行される

// twitterの画像を表示したときのC-sを拡張
// 画像のファイル名を「～.jpg-orig」「～.png-orig」ではなく「～-orig.jpg」「～-orig.png」にする

import { INITIAL_OPTIONS, STRIP_IMAGE_SUFFIX } from './helpers/Constants';
import { updateOptions, collectUrlParams } from './helpers/Utils';

const options = { ...INITIAL_OPTIONS };

const getImageFilenameByUrl = (imgUrl: string) => {
  const params = collectUrlParams(imgUrl);
  if (!params) {
    return null;
  }

  const { pathname, format, name } = params;
  const basename = pathname.match(/([^/.]*?)(?:\..+)?$/)![1];

  return `${basename}${name ? `-${name}` : ''}.${format}`;
};

document.addEventListener('DOMContentLoaded', () => {
  updateOptions(options);
});

// キーを押したとき
document.addEventListener('keydown', e => {
  // if 設定が有効なら
  // かつ 押されたキーがC-s の状態なら
  // かつ 開いているURLが画像URLの定形なら(pbs.twimg.comを使うものは他にも存在するので)
  if (
    options[STRIP_IMAGE_SUFFIX] !== 'isfalse' &&
    e.key === 's' &&
    (e.ctrlKey || e.metaKey)
  ) {
    const imageSrc = document.querySelector('img')!.src;
    const filename = getImageFilenameByUrl(imageSrc);
    if (!filename) {
      return;
    }
    // もとの挙動(ブラウザが行う保存)をしないよう中止
    e.preventDefault();

    // download属性に正しい拡張子の画像名を入れたaタグをつくってクリックする
    const a = document.createElement('a');
    a.href = window.location.href;
    a.setAttribute('download', filename);
    a.dispatchEvent(new MouseEvent('click'));
  }
});
