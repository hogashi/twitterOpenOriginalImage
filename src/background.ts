import { GET_LOCAL_STORAGE } from './main';

// バックグラウンドで実行される

// chrome.runtime.sendMessage で送るメッセージ
export interface MessageRequest {
  method: string;
}
// chrome.runtime.sendMessage で返るメッセージ
export interface MessageResponse {
  data: { [key: string]: string } | null;
}

chrome.runtime.onMessage.addListener(
  (
    request: MessageRequest,
    _,
    sendResponse: (res: MessageResponse) => void
  ) => {
    // console.log(chrome.runtime.lastError);
    if (request.method === GET_LOCAL_STORAGE) {
      sendResponse({ data: localStorage });
    } else {
      sendResponse({ data: null });
    }
    return true;
  }
);
