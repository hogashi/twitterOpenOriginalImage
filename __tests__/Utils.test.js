import * as main from '../src/main';
const {
  isTrue,
  isFalse,
  OPTION_KEYS,
  printException,
  collectUrlParams,
  formatUrl,
  openImages,
  getOptions,
  setStyle,
  onOriginalButtonClick,
  getButtonSetter,
  ButtonSetter,
  ButtonSetterTweetDeck,
} = main;

const makeResultParams = ({ format, name }) => {
  return {
    protocol: 'https:',
    host: 'pbs.twimg.com',
    pathname: '/media/hogefuga123',
    format,
    name: name || null,
  };
};

describe('Utils', () => {
  describe('printException', () => {
    it('エラーメッセージの表示(予期せぬ状況の確認)', () => {
      console.log = jest.fn();
      printException('exception message');
      expect(console.log.mock.calls[0][0]).toBe(
        'tooi: exception message at: http://localhost/'
      );
    });
  });

  describe('URLのツール', () => {
    // URLの変換のテストの場合たち
    const cases = [
      {
        title: '何もなし',
        url: 'https://pbs.twimg.com/media/hogefuga123',
        params: { format: 'jpg' },
      },
      {
        title: '.jpg',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg',
        params: { format: 'jpg' },
      },
      {
        title: '.png',
        url: 'https://pbs.twimg.com/media/hogefuga123.png',
        params: { format: 'png' },
      },
      {
        title: '.jpg:orig',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg:orig',
        params: { format: 'jpg', name: 'orig' },
      },
      {
        title: '.jpg:large',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg:large',
        params: { format: 'jpg', name: 'large' },
      },
      {
        title: '?format=jpg',
        url: 'https://pbs.twimg.com/media/hogefuga123?format=jpg',
        params: { format: 'jpg' },
      },
      {
        title: '?format=png',
        url: 'https://pbs.twimg.com/media/hogefuga123?format=png',
        params: { format: 'png' },
      },
      {
        title: '.jpg?format=jpg',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg?format=jpg',
        params: { format: 'jpg' },
      },
      {
        title: '.jpg?format=png',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg?format=png',
        params: { format: 'jpg' },
      },
      {
        title: '.png?format=jpg',
        url: 'https://pbs.twimg.com/media/hogefuga123.png?format=jpg',
        params: { format: 'png' },
      },
      {
        title: '.jpg?name=large',
        url: 'https://pbs.twimg.com/media/hogefuga123.jpg?name=large',
        params: { format: 'jpg', name: 'large' },
      },
      {
        title: '?format=jpg&name=large',
        url: 'https://pbs.twimg.com/media/hogefuga123?format=jpg&name=large',
        params: { format: 'jpg', name: 'large' },
      },
      {
        title: '.png?format=jpg&name=orig',
        url: 'https://pbs.twimg.com/media/hogefuga123.png?format=jpg&name=orig',
        params: { format: 'png', name: 'orig' },
      },
    ];

    describe('collectUrlParams 画像urlの要素を集める', () => {
      cases.forEach(singleCase => {
        const { title, url, params } = singleCase;
        it(title, () => {
          expect(collectUrlParams(url)).toStrictEqual(makeResultParams(params));
        });
      });

      it('twitterの画像URLでないときnull', () => {
        expect(collectUrlParams('https://twitter.com/tos')).toBe(null);
      });
    });

    describe('formatUrl 画像URLを https～?format=〜&name=orig に揃える', () => {
      cases.forEach(singleCase => {
        const { title, url, params } = singleCase;
        it(title, () => {
          expect(formatUrl(url)).toBe(
            `https://pbs.twimg.com/media/hogefuga123?format=${params.format}&name=orig`
          );
        });
      });

      it('twitterの画像URLでないときそのまま', () => {
        expect(formatUrl('https://twitter.com/tos')).toBe(
          'https://twitter.com/tos'
        );
      });
    });
  });

  describe('openImages 画像を開く', () => {
    it('画像URLを1つ渡したとき開く', () => {
      window.open = jest.fn();
      openImages(['https://pbs.twimg.com/media/1st?format=jpg&name=orig']);
      expect(window.open.mock.calls.length).toBe(1);
      expect(window.open.mock.calls[0][0]).toBe(
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig'
      );
    });

    it('画像URLを2つ渡したとき逆順に開く', () => {
      window.open = jest.fn();
      openImages([
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig',
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig',
      ]);
      expect(window.open.mock.calls.length).toBe(2);
      expect(window.open.mock.calls[0][0]).toBe(
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig'
      );
      expect(window.open.mock.calls[1][0]).toBe(
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig'
      );
    });

    it('画像URLを4つ渡したとき逆順に開く', () => {
      window.open = jest.fn();
      openImages([
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig',
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig',
        'https://pbs.twimg.com/media/3rd?format=jpg&name=orig',
        'https://pbs.twimg.com/media/4th?format=jpg&name=orig',
      ]);
      expect(window.open.mock.calls.length).toBe(4);
      expect(window.open.mock.calls[0][0]).toBe(
        'https://pbs.twimg.com/media/4th?format=jpg&name=orig'
      );
      expect(window.open.mock.calls[1][0]).toBe(
        'https://pbs.twimg.com/media/3rd?format=jpg&name=orig'
      );
      expect(window.open.mock.calls[2][0]).toBe(
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig'
      );
      expect(window.open.mock.calls[3][0]).toBe(
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig'
      );
    });

    it('画像URLでないURLを1つ渡したときもそのまま開く', () => {
      window.open = jest.fn();
      openImages(['https://twitter.com/tos']);
      expect(window.open.mock.calls.length).toBe(1);
      expect(window.open.mock.calls[0][0]).toBe('https://twitter.com/tos');
    });

    it('undefinedを1つ渡したとき開かない', () => {
      window.open = jest.fn();
      openImages([undefined]);
      expect(window.open.mock.calls.length).toBe(0);
    });

    it('URLとundefinedを混ぜたときURLだけ開いてundefinedは開かない', () => {
      window.open = jest.fn();
      openImages([
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig',
        undefined,
        'https://twitter.com/tos',
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig',
      ]);
      expect(window.open.mock.calls.length).toBe(3);
      expect(window.open.mock.calls[0][0]).toBe(
        'https://pbs.twimg.com/media/2nd?format=jpg&name=orig'
      );
      expect(window.open.mock.calls[1][0]).toBe('https://twitter.com/tos');
      expect(window.open.mock.calls[2][0]).toBe(
        'https://pbs.twimg.com/media/1st?format=jpg&name=orig'
      );
    });

    it('要素0個の配列を渡したとき開かない', () => {
      window.open = jest.fn();
      openImages([]);
      expect(window.open.mock.calls.length).toBe(0);
    });
  });

  describe('getOptions', () => {
    describe('Chrome拡張機能のとき', () => {
      const originalChrome = window.chrome;
      beforeAll(() => {
        delete window.chrome;
        window.chrome = { runtime: { id: 'id' } };
      });
      afterAll(() => {
        window.chrome = originalChrome;
      });

      it('初期設定を取得できる', () => {
        const expected = {};
        OPTION_KEYS.forEach(key => {
          expected[key] = isTrue;
        });
        window.chrome.runtime.sendMessage = jest.fn((_, callback) =>
          callback({ data: {} })
        );
        expect(getOptions()).resolves.toStrictEqual(expected);
      });

      it('設定した値を取得できる', () => {
        const expected = {};
        OPTION_KEYS.forEach((key, i) => {
          expected[key] = i % 2 === 0 ? isTrue : isFalse;
        });
        window.chrome.runtime.sendMessage = jest.fn((_, callback) =>
          callback({ data: { ...expected } })
        );
        return getOptions().then(options => {
          return expect(options).toStrictEqual(expected);
        });
      });

      it('設定が取得できなかったらreject', () => {
        const expected = {};
        OPTION_KEYS.forEach(key => {
          expected[key] = isTrue;
        });
        window.chrome.runtime.sendMessage = jest.fn((_, callback) =>
          callback({})
        );
        expect(getOptions()).rejects.toBeUndefined();
      });
    });

    describe('Chrome拡張機能でないとき', () => {
      const originalChrome = window.chrome;
      beforeAll(() => {
        delete window.chrome;
        window.chrome = undefined;
      });
      afterAll(() => {
        window.chrome = originalChrome;
      });

      it('初期設定を取得できる', () => {
        const expected = {};
        OPTION_KEYS.forEach(key => {
          expected[key] = isFalse;
        });
        expect(getOptions()).resolves.toStrictEqual(expected);
      });
    });
  });

  describe('setStyle DOM要素にスタイルを当てる', () => {
    it('スタイル当たる', () => {
      const div = document.createElement('div');
      expect(div).toMatchSnapshot();
      setStyle(div, {
        display: 'none',
        color: '#123',
        'background-color': 'rgba(12, 34, 56, 0.7)',
      });
      expect(div).toMatchSnapshot();
    });
    it('空のスタイル渡すと何もしない', () => {
      const div = document.createElement('div');
      expect(div).toMatchSnapshot();
      setStyle(div, {});
      expect(div).toMatchSnapshot();
    });
    it('すでにあるスタイルは上書きされるが消えることはない', () => {
      const div = document.createElement('div');
      div.style.color = '#456';
      div.style.backgroundColor = '#789abc';
      div.style.fontSize = '150px';
      expect(div).toMatchSnapshot();
      setStyle(div, {
        display: 'none',
        color: '#123',
        'background-color': 'rgba(12, 34, 56, 0.7)',
      });
      expect(div).toMatchSnapshot();
    });
  });

  describe('onOriginalButtonClick ボタンがクリックされたときのコールバック', () => {
    it('イベントを止める', () => {
      /* SKIP: なぜかうまくmockできないので飛ばす */
      // jest.spyOn(main, 'openImages');

      const event = {
        preventDefault: jest.fn(),
        stopPropagation: jest.fn(),
      };
      const imgSrcs = ['src1', 'src2'];
      onOriginalButtonClick(event, imgSrcs);

      expect(event.preventDefault).toHaveBeenCalledTimes(1);
      expect(event.stopPropagation).toHaveBeenCalledTimes(1);

      // expect(openImages).toHaveBeenCalledTimes(1);
      // expect(openImages.mock.calls[0][0]).toStrictEqual(imgSrcs);
    });
  });

  describe('getButtonSetter ボタン設置するクラスのゲッタ', () => {
    const originalLocation = window.location;
    beforeAll(() => {
      delete window.location;
    });
    afterAll(() => {
      window.location = originalLocation;
    });

    it('公式WebではButtonSetter', () => {
      window.location = new URL('https://twitter.com');
      expect(getButtonSetter()).toBeInstanceOf(ButtonSetter);
    });
    it('TweetDeckではButtonSetterTweetDeck', () => {
      window.location = new URL('https://tweetdeck.twitter.com');
      expect(getButtonSetter()).toBeInstanceOf(ButtonSetterTweetDeck);
    });
    it('どちらでもなかったらButtonSetter', () => {
      window.location = new URL('https://hoge.test');
      expect(getButtonSetter()).toBeInstanceOf(ButtonSetter);
    });
  });
});
