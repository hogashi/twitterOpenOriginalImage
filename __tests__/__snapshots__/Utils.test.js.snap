// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utils setStyle DOM要素にスタイルを当てる すでにあるスタイルは上書きされるが消えることはない 1`] = `
<div
  style="color: rgb(68, 85, 102); background-color: rgb(120, 154, 188); font-size: 150px;"
/>
`;

exports[`Utils setStyle DOM要素にスタイルを当てる すでにあるスタイルは上書きされるが消えることはない 2`] = `
<div
  style="color: rgb(17, 34, 51); background-color: rgba(12, 34, 56, 0.7); font-size: 150px; display: none;"
/>
`;

exports[`Utils setStyle DOM要素にスタイルを当てる スタイル当たる 1`] = `<div />`;

exports[`Utils setStyle DOM要素にスタイルを当てる スタイル当たる 2`] = `
<div
  style="display: none; color: rgb(17, 34, 51); background-color: rgba(12, 34, 56, 0.7);"
/>
`;

exports[`Utils setStyle DOM要素にスタイルを当てる 空のスタイル渡すと何もしない 1`] = `<div />`;

exports[`Utils setStyle DOM要素にスタイルを当てる 空のスタイル渡すと何もしない 2`] = `<div />`;
